flink-runtime/src/test/java/org/apache/flink/runtime/concurrent/ScheduledFutureAdapterTest.java

```diff
@@ -0,0 +1,139 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.concurrent;
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+import java.util.concurrent.Delayed;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static org.powermock.api.mockito.PowerMockito.spy;
+
+/**
+ * Unit tests for {@link ScheduledFutureAdapter}.
+ */
+public class ScheduledFutureAdapterTest {
+
+	private ScheduledFutureAdapter objectUnderTest;
+	private Future<Integer> innerDelegate;
+	private AtomicBoolean resultCancel;
+	private AtomicBoolean resultIsCancelled;
+	private AtomicBoolean resultIsDone;
+
+	@Before
+	public void before() throws Exception {
+		this.resultCancel = new AtomicBoolean(false);
+		this.resultIsCancelled = new AtomicBoolean(false);
+		this.resultIsDone = new AtomicBoolean(false);
+
+		this.innerDelegate = spy(new Future<Integer>() {
+			@Override
+			public boolean cancel(boolean mayInterruptIfRunning) {
+				return resultCancel.get();
+			}
+
+			@Override
+			public boolean isCancelled() {
+				return resultIsCancelled.get();
+			}
+
+			@Override
+			public boolean isDone() {
+				return resultIsDone.get();
+			}
+
+			@Override
+			public Integer get() {
+				return 4711;
+			}
+
+			@Override
+			public Integer get(long timeout, TimeUnit unit) {
+				return 4711;
+			}
+		});
+
+		this.objectUnderTest = new ScheduledFutureAdapter<>(innerDelegate, 4200L, TimeUnit.MILLISECONDS);
+	}
+
+	@Test
+	public void testForwardedMethods() throws Exception {
+
+		Assert.assertEquals(4711, objectUnderTest.get());
+		Mockito.verify(innerDelegate).get();
+
+		Assert.assertEquals(4711, objectUnderTest.get(42L, TimeUnit.SECONDS));
+		Mockito.verify(innerDelegate).get(42L, TimeUnit.SECONDS);
+
+		Assert.assertEquals(resultCancel.get(), objectUnderTest.cancel(true));
+		Mockito.verify(innerDelegate, Mockito.times(1)).cancel(true);
+
+		resultCancel.set(!resultCancel.get());
+		Assert.assertEquals(resultCancel.get(), objectUnderTest.cancel(true));
+		Mockito.verify(innerDelegate, Mockito.times(2)).cancel(true);
+
+		Assert.assertEquals(resultIsCancelled.get(), objectUnderTest.isCancelled());
+		Mockito.verify(innerDelegate, Mockito.times(1)).isCancelled();
+
+		resultIsCancelled.set(!resultIsCancelled.get());
+		Assert.assertEquals(resultIsCancelled.get(), objectUnderTest.isCancelled());
+		Mockito.verify(innerDelegate, Mockito.times(2)).isCancelled();
+
+		Assert.assertEquals(resultIsDone.get(), objectUnderTest.isDone());
+		Mockito.verify(innerDelegate, Mockito.times(1)).isDone();
+
+		resultIsDone.set(!resultIsDone.get());
+		Assert.assertEquals(resultIsDone.get(), objectUnderTest.isDone());
+		Mockito.verify(innerDelegate, Mockito.times(2)).isDone();
```

Comment:
Here in this case it probably does not make much of a difference but I think we could easily do all the checks without Mockito by simply having method counters. Mockito adds a bit of complexity which usually makes tests slower to execute.

Reply:
I always think this is one of the few cases where mockito is not a supid thing to use.Do you think it is so much slower that we rather write more boilerplate code?

Reply:
Anyways, as there is already an adhock impl there, it might really be ok to introduce counts.